File: pkg/agent/mdns/publisher.go
/*
 * Copyright 2025 Carver Automation Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package mdns

import (
	"context"
	"fmt"
	"time"

	"github.com/nats-io/nats.go"
	"github.com/nats-io/nats.go/jetstream"

	"github.com/carverauto/serviceradar/pkg/logger"
	"github.com/carverauto/serviceradar/pkg/natsutil"
)

// Publisher reads protobuf-encoded mDNS records from a channel and publishes
// them to NATS JetStream.
type Publisher struct {
	config *Config
	ch     <-chan []byte
	nc     *nats.Conn
	js     jetstream.JetStream
	logger logger.Logger
}

// NewPublisher creates a new NATS JetStream publisher.
func NewPublisher(config *Config, ch <-chan []byte, log logger.Logger) *Publisher {
	return &Publisher{
		config: config,
		ch:     ch,
		logger: log,
	}
}

// Connect establishes the NATS connection with exponential backoff retry.
func (p *Publisher) Connect(ctx context.Context) error {
	var (
		attempt        uint32
		backoff        = 500 * time.Millisecond
		maxBackoff     = 30 * time.Second
		maxAttempts    uint32 = 60
	)

	for {
		attempt++
		err := p.connectOnce(ctx)
		if err == nil {
			return nil
		}

		if attempt >= maxAttempts {
			return fmt.Errorf("%w: %v (after %d attempts)", ErrNATSConnection, err, maxAttempts)
		}

		p.logger.Warn().
			Err(err).
			Uint32("attempt", attempt).
			Dur("backoff", backoff).
			Msg("NATS connection attempt failed, retrying")

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(backoff):
		}

		backoff *= 2
		if backoff > maxBackoff {
			backoff = maxBackoff
		}
	}
}

func (p *Publisher) connectOnce(ctx context.Context) error {
	var opts []nats.Option

	if p.config.NATSCredsFile != "" {
		opts = append(opts, nats.UserCredentials(p.config.NATSCredsFile))
	}

	nc, err := natsutil.ConnectWithSecurity(ctx, p.config.NATSUrl, p.config.Security, opts...)
	if err != nil {
		return err
	}

	js, err := jetstream.New(nc)
	if err != nil {
		nc.Close()
		return fmt.Errorf("failed to create JetStream context: %w", err)
	}

	// Ensure stream exists
	if err := p.ensureStream(ctx, js); err != nil {
		nc.Close()
		return err
	}

	p.nc = nc
	p.js = js

	p.logger.Info().
		Str("url", p.config.NATSUrl).
		Str("stream", p.config.StreamName).
		Msg("Connected to NATS and ensured stream exists")

	return nil
}

func (p *Publisher) ensureStream(ctx context.Context, js jetstream.JetStream) error {
	requiredSubjects := p.config.StreamSubjectsResolved()

	stream, err := js.Stream(ctx, p.config.StreamName)
	if err != nil {
		// Stream doesn't exist, create it
		_, createErr := js.CreateOrUpdateStream(ctx, jetstream.StreamConfig{
			Name:     p.config.StreamName,
			Subjects: requiredSubjects,
			Storage:  jetstream.FileStorage,
			MaxBytes: p.config.StreamMaxBytes,
			MaxAge:   24 * time.Hour,
		})
		return createErr
	}

	// Stream exists, check if subjects need updating
	info, err := stream.Info(ctx)
	if err != nil {
		return fmt.Errorf("failed to get stream info: %w", err)
	}

	needsUpdate := false
	currentSubjects := info.Config.Subjects
	for _, required := range requiredSubjects {
		found := false
		for _, existing := range currentSubjects {
			if existing == required {
				found = true
				break
			}
		}
		if !found {
			currentSubjects = append(currentSubjects, required)
			needsUpdate = true
		}
	}

	if needsUpdate {
		cfg := info.Config
		cfg.Subjects = currentSubjects
		_, err = js.CreateOrUpdateStream(ctx, cfg)
		if err != nil {
			return fmt.Errorf("failed to update stream subjects: %w", err)
		}
	}

	return nil
}

// Run reads from the channel and publishes batches to NATS.
// Blocks until the channel is closed or context is cancelled.
func (p *Publisher) Run(ctx context.Context) {
	batch := make([][]byte, 0, p.config.BatchSize)
	timeout := time.Duration(p.config.PublishTimeoutMs) * time.Millisecond

	p.logger.Info().
		Str("subject", p.config.Subject).
		Msg("mDNS publisher started")

	for {
		select {
		case <-ctx.Done():
			if len(batch) > 0 {
				p.publishBatch(ctx, batch, timeout)
			}
			return
		case msg, ok := <-p.ch:
			if !ok {
				if len(batch) > 0 {
					p.publishBatch(ctx, batch, timeout)
				}
				p.logger.Info().Msg("mDNS publisher channel closed, shutting down")
				return
			}

			batch = append(batch, msg)

			// Drain any immediately-available messages to build a batch
			draining := true
			for draining && len(batch) < p.config.BatchSize {
				select {
				case msg, ok := <-p.ch:
					if !ok {
						draining = false
					} else {
						batch = append(batch, msg)
					}
				default:
					draining = false
				}
			}

			if len(batch) > 0 {
				p.publishBatch(ctx, batch, timeout)
				batch = batch[:0]
			}
		}
	}
}

func (p *Publisher) publishBatch(ctx context.Context, batch [][]byte, timeout time.Duration) {
	pubCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	for _, msg := range batch {
		_, err := p.js.Publish(pubCtx, p.config.Subject, msg)
		if err != nil {
			p.logger.Error().Err(err).Msg("Failed to publish mDNS record to NATS")
		}
	}
}

// Close closes the NATS connection.
func (p *Publisher) Close() {
	if p.nc != nil {
		p.nc.Close()
	}
}

File: rust/netflow-collector/src/publisher.rs
use crate::config::{Config, SecurityMode};
use anyhow::{Context, Result};
use async_nats::jetstream::{self, stream::StorageType};
use async_nats::{Client, ConnectOptions};
use log::{error, info, warn};
use std::cmp::min;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::mpsc;
use tokio::sync::mpsc::error::TryRecvError;
use tokio::time::{sleep, timeout};

pub struct Publisher {
    config: Arc<Config>,
    rx: mpsc::Receiver<Vec<u8>>,
}

impl Publisher {
    pub fn new(config: Arc<Config>, rx: mpsc::Receiver<Vec<u8>>) -> Self {
        Self { config, rx }
    }

    pub async fn run(mut self) -> Result<()> {
        let (_, js) = self.connect_with_retry().await?;

        let mut batch = Vec::with_capacity(self.config.batch_size);
        let timeout_duration = Duration::from_millis(self.config.publish_timeout_ms);

        info!(
            "Publisher started, publishing to subject: {}",
            self.config.subject
        );

        loop {
            // Receive messages from the listener channel
            let msg = match self.rx.recv().await {
                Some(msg) => msg,
                None => {
                    // Channel closed, publish remaining messages and exit
                    if !batch.is_empty() {
                        self.publish_batch(&js, &mut batch, timeout_duration).await;
                    }
                    info!("Publisher channel closed, shutting down");
                    return Ok(());
                }
            };

            batch.push(msg);

            // Drain any immediately-available messages to build a batch
            let mut closed = false;
            while batch.len() < self.config.batch_size {
                match self.rx.try_recv() {
                    Ok(msg) => batch.push(msg),
                    Err(TryRecvError::Empty) => break,
                    Err(TryRecvError::Disconnected) => {
                        closed = true;
                        break;
                    }
                }
            }

            // Publish whatever we have so low-volume streams still emit data
            if !batch.is_empty() {
                self.publish_batch(&js, &mut batch, timeout_duration).await;
            }

            if closed {
                info!("Publisher channel closed, shutting down");
                return Ok(());
            }
        }
    }

    async fn publish_batch(
        &self,
        js: &jetstream::Context,
        batch: &mut Vec<Vec<u8>>,
        timeout_duration: Duration,
    ) {
        for msg in batch.drain(..) {
            match js.publish(self.config.subject.clone(), msg.into()).await {
                Ok(ack) => {
                    // Wait for ack with timeout
                    if timeout(timeout_duration, ack).await.is_err() {
                        warn!("NATS ack timed out after {:?}", timeout_duration);
                        // TODO: Increment metrics counter for timeouts
                    }
                }
                Err(e) => {
                    error!("Failed to publish to NATS: {}", e);
                    // TODO: Increment metrics counter for publish failures
                }
            }
        }
    }

    async fn connect_once(&self) -> Result<(Client, jetstream::Context)> {
        let mut options = ConnectOptions::new();

        // Apply security configuration
        if let Some(sec) = &self.config.security {
            match sec.mode {
                SecurityMode::Mtls => {
                    // Apply CA file if provided
                    if let Some(ca_path) = sec.ca_file_path() {
                        options = options.add_root_certificates(ca_path);
                    }

                    // Apply client certificate and key for mTLS
                    if let (Some(cert_path), Some(key_path)) =
                        (sec.cert_file_path(), sec.key_file_path())
                    {
                        options = options.add_client_certificate(cert_path, key_path);
                    }
                }
                SecurityMode::None => {
                    // No TLS
                }
            }
        }

        if let Some(creds_file) = &self.config.nats_creds_file {
            options = options
                .credentials_file(creds_file)
                .await
                .with_context(|| format!("Failed to load NATS creds file {}", creds_file))?;
        }

        // Connect to NATS server
        let client = options.connect(&self.config.nats_url).await?;
        let js = jetstream::new(client.clone());

        // Ensure the target stream exists and includes all required subjects
        let required_subjects = self.config.stream_subjects_resolved();
        match js.get_stream(&self.config.stream_name).await {
            Ok(mut existing_stream) => {
                let info = existing_stream.info().await?;
                let mut current_subjects = info.config.subjects.clone();
                let mut needs_update = false;

                for required in &required_subjects {
                    if !current_subjects.contains(required) {
                        current_subjects.push(required.clone());
                        needs_update = true;
                    }
                }

                if needs_update {
                    let mut updated_config = info.config.clone();
                    updated_config.subjects = current_subjects;
                    js.update_stream(updated_config).await?;
                    js.get_stream(&self.config.stream_name).await?;
                }
            }
            Err(_) => {
                let stream_config = jetstream::stream::Config {
                    name: self.config.stream_name.clone(),
                    subjects: required_subjects.clone(),
                    storage: StorageType::File,
                    max_bytes: self.config.stream_max_bytes,
                    max_age: Duration::from_secs(24 * 60 * 60), // 24 hours
                    ..Default::default()
                };
                js.get_or_create_stream(stream_config).await?;
            }
        }

        info!(
            "Connected to NATS at {} and ensured stream '{}' exists",
            self.config.nats_url, self.config.stream_name
        );

        Ok((client, js))
    }

    async fn connect_with_retry(&self) -> Result<(Client, jetstream::Context)> {
        let mut attempt: u32 = 0;
        let initial_backoff = Duration::from_millis(500);
        let max_backoff = Duration::from_secs(30);
        let mut backoff = min(initial_backoff, max_backoff);
        let max_attempts = 60; // Retry for ~30 minutes with exponential backoff

        loop {
            attempt += 1;
            match self.connect_once().await {
                Ok(conn) => return Ok(conn),
                Err(err) => {
                    if attempt >= max_attempts {
                        error!(
                            "NATS connection attempt {} failed: {}. Giving up after {} attempts.",
                            attempt, err, max_attempts
                        );
                        return Err(err);
                    }

                    warn!(
                        "NATS connection attempt {} failed: {}. Retrying in {:?}...",
                        attempt, err, backoff
                    );
                    sleep(backoff).await;

                    // Exponential backoff
                    let doubled = backoff.checked_mul(2).unwrap_or(max_backoff);
                    backoff = min(doubled, max_backoff);
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_publisher_creation() {
        let config = Arc::new(Config {
            listen_addr: "0.0.0.0:2055".to_string(),
            buffer_size: 65536,
            nats_url: "nats://localhost:4222".to_string(),
            nats_creds_file: None,
            stream_name: "flows".to_string(),
            subject: "flows.raw.netflow".to_string(),
            stream_subjects: None,
            stream_max_bytes: 10 * 1024 * 1024 * 1024,
            partition: "default".to_string(),
            max_templates: 2000,
            max_template_fields: 10_000,
            channel_size: 1000,
            batch_size: 100,
            publish_timeout_ms: 5000,
            drop_policy: crate::config::DropPolicy::DropOldest,
            security: None,
            pending_flows: None,
            metrics_addr: None,
        });

        let (_tx, rx) = mpsc::channel(1000);
        let publisher = Publisher::new(config, rx);
        assert!(publisher.config.subject == "flows.raw.netflow");
    }
}
